lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,256 @@
+import 'dart:collection';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+Element _getLeftElement(AssignmentExpression assignment) {
+  final leftPart = assignment.leftHandSide;
+  return leftPart is SimpleIdentifier
+      ? DartTypeUtilities.getCanonicalElement(leftPart.bestElement)
+      : leftPart is PropertyAccess
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isBreakStatement(Statement statement) => (statement is BreakStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is BreakStatement));
+
+bool _isContinueStatement(
+        Statement statement) =>
+    (statement is ContinueStatement ||
+        (statement is Block &&
+            statement.statements.length == 1 &&
+            statement.statements.first is ContinueStatement));
+
+bool _isReturnStatement(Statement statement) => (statement is ReturnStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is ReturnStatement));
+
+abstract class ConditionScopeVisitor extends RecursiveAstVisitor {
+  final Queue<Queue<_ExpressionBox>> environments = new Queue();
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
```

Comment:
Yes but, in that case, we would add to the scope something like, null is undefined, and that would not change the behavior of invariant booleans, at least we do something like:

```dart
if (notDefined2 < 2) {
  notDefined = 3; // we undefine null
  if (notDefined2 < 2) { // this won't be linted, because notDefined is null and null was undefined just before this condition is evaluated
  }
}
```
My point is that checking those nulls wont change the behavior in code with not null elements in conditions inside if/while/do/for, so we should take care about conditions with null elements inside too, what do you think?